lib/rdf-model/src/alloc/heap_term.rs

```diff
@@ -93,3 +90,64 @@ impl From<String> for HeapTerm {
         Self::literal(&value)
     }
 }
+
+#[cfg(feature = "borsh")]
+impl borsh::BorshSerialize for HeapTerm {
+    fn serialize<W: borsh::io::Write>(&self, writer: &mut W) -> borsh::io::Result<()> {
+        match self {
+            HeapTerm::Iri(value) => {
+                0x01u8.serialize(writer)?;
+                value.serialize(writer)
+            }
+            HeapTerm::BNode(value) => {
+                0x02u8.serialize(writer)?;
+                value.serialize(writer)
+            }
+            HeapTerm::Literal(value) => {
+                0x03u8.serialize(writer)?;
+                value.serialize(writer)
+            }
+            HeapTerm::LiteralWithDatatype(value, datatype) => {
+                0x04u8.serialize(writer)?;
+                value.serialize(writer)?;
+                datatype.serialize(writer)
+            }
+            HeapTerm::LiteralWithLanguage(value, language) => {
+                0x05u8.serialize(writer)?;
+                value.serialize(writer)?;
+                language.serialize(writer)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "borsh")]
+impl borsh::BorshDeserialize for HeapTerm {
+    fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
+        Ok(match u8::deserialize_reader(reader)? {
+            0x01 => {
+                let value = String::deserialize_reader(reader)?;
+                HeapTerm::Iri(value)
+            }
+            0x02 => {
+                let value = String::deserialize_reader(reader)?;
+                HeapTerm::BNode(value)
+            }
+            0x03 => {
+                let value = String::deserialize_reader(reader)?;
+                HeapTerm::Literal(value)
+            }
+            0x04 => {
+                let value = String::deserialize_reader(reader)?;
+                let datatype = String::deserialize_reader(reader)?;
+                HeapTerm::LiteralWithDatatype(value, datatype)
+            }
+            0x05 => {
+                let value = String::deserialize_reader(reader)?;
+                let language = String::deserialize_reader(reader)?;
+                HeapTerm::LiteralWithLanguage(value, language)
+            }
+            _ => return Err(borsh::io::ErrorKind::InvalidData.into()),
+        })
+    }
+}
```

Comment:
Ideally, this code wouldn't be in rdf-model, but rather in rdf-borsh. Because there is not only "one true way" to (de)serialize RDF terms in Borsh, unlike with Serde where that's a more likely possibility.

Almost the whole reason for why the `BorshTerm` struct exists is so that the Borsh traits could be implemented for it in rdf-borsh.